### PR TITLE
[SPARK-43504][K8S] Mounts the hadoop config map on the executor pod

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?

In this pr, for spark on k8s, the hadoop config map will be mounted in executor side as well.
Before, the  hadoop config map is only mounted in driver side.
### Why are the changes needed?

Since [SPARK-25815](https://issues.apache.org/jira/browse/SPARK-25815) [,](https://github.com/apache/spark/pull/22911,) the hadoop config map will not be mounted in executor side.

Per the  https://github.com/apache/spark/pull/22911 description:

> The main two things that don't need to happen in executors anymore are:
> 1. adding the Hadoop config to the executor pods: this is not needed
> since the Spark driver will serialize the Hadoop config and send
> it to executors when running tasks. 

But in fact, the executor still need the hadoop configuration.

image

As shown in above picture, the driver can resolve `hdfs://zeus`, but the executor can not.

so we still need to mount the hadoop config map in executor side.

### Does this PR introduce _any_ user-facing change?

Yes, users do not need to take workarounds to make executors load the hadoop configuration.
Such as:
- including hadoop conf in executor image
- placing hadoop conf files under `SPARK_CONF_DIR`.
### How was this patch tested?

UT.

### CODE REVIEW

**Comment by pan3793** [comment]
I encountered the same issues w/ Spark 3.3.1. This sounds like a regression (I suppose it works before SPARK-25815, I don't have experience w/ running such an old version of Spark on K8s).

The key point is that the executor needs to download artifacts during the bootstrap phase, so the assumption in SPARK-25815 is not always true.

> adding the Hadoop config to the executor pods: this is not needed
> since the Spark driver will serialize the Hadoop config and send
> it to executors when running tasks.

Given the executor use `SparkHadoopUtil.get.newConfiguration(conf)` to construct Hadoop conf,  we can put the related hdfs/s3 configurations into `spark-defaults.conf` w/ `spark.hadoop.` prefix as a workaround.

https://github.com/apache/spark/blob/0df4c01b7c4d4476fe0de9dccb3425cc1295fc85/core/src/main/scala/org/apache/spark/executor/Executor.scala#L1006-L1012

This PR definitely fixes some use cases, @turboFei would you mind updating "Does this PR introduce any user-facing change?"

**Comment by turboFei** [comment]
> would you mind updating "Does this PR introduce any user-facing change?"

updated

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStep.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.spark.deploy.k8s.Constants._
+
+/**
+ * Mounts the Hadoop configuration on the executor pod.
+ */
+private[spark] class HadoopConfExecutorFeatureStep(conf: KubernetesConf)
+  extends KubernetesFeatureConfigStep {
+
+  override def configurePod(pod: SparkPod): SparkPod = {
+    conf.getOption(HADOOP_CONFIG_MAP_NAME) match {
```
[the remark this answers is not in the document]
Comment (turboFei) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStep.scala:33]: thanks, will check it

**Comment by turboFei** [comment]
seems the k8s integration testing is stuck, will check this pr in our dev hadoop cluster tomorrow.

**Review by dongjoon-hyun — DISMISSED** [review]
Thank you for making a PR, @turboFei .

However, this PR might cause a outage because the number of configMap is controlled by quota.
```
$ kubectl describe quota | grep configmaps
count/configmaps                                                  4     150
```

To avoid the production outage, this should be under a new configuration with `false` by default at least.

**Comment by turboFei** [comment]
thanks for comments, I will check it

**Comment by turboFei** [comment]
> this PR doesn't create new ConfigMaps, it either uses a user pre-set config map (no creation) or just reuse the config map created by driver which is created if necessary.

yes, this PR doesn't create new ConfigMap.

**Comment by dongjoon-hyun** [comment]
Oh, got it. Thank you for correcting me.

**Comment by turboFei** [comment]
the UT has passed, gentle ping @dongjoon-hyun

**Review by advancedxy — APPROVED** [review]
lgtm.
left one minor comment, I'm ok with merging this as it is.

**Comment by turboFei** [comment]
thanks @advancedxy add more UT to check additionalPodSystemProperties and executor pod.

**Review by advancedxy — APPROVED** [review]

**Comment by turboFei** [comment]
gentle ping @dongjoon-hyun would you like to review again? thanks

**Comment by yaooqinn** [comment]
The Hadoop configurations can be propagated after https://github.com/apache/spark/pull/27735. And putting and locating extra configuration files in SPARK_HOME/conf is also a suggested way from our docs, so is this step necessary?

Alternatively, if both exist, what is the precedence between them? Is it idempotent?

**Comment by turboFei** [comment]
> And putting and locating extra configuration files in SPARK_HOME/conf is also a suggested way from our docs, so is this step necessary?

I think it is necessary.

Hadoop and spark are different components, it is better to maintain them separately.

In our company, we have conf version for hadoop conf, so we do not put hadoop config files under SPARK_HOME/conf and use soft link to manage the hadoop conf.


> Alternatively, if both exist, what is the precedence between them? Is it idempotent?

In this pr, it just mounts the hadoop config map in the executor side and the hadoop conf mounted is absolute same with that in driver pod.

As shown below, the SPARK_CONF_DIR has higher precedence.
https://github.com/apache/spark/blob/e096bce604ed6fab37713437ac0d985673910537/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh#L68-L76

**Comment by turboFei** [comment]
gentle ping @yaooqinn @dongjoon-hyun

**Comment by yaooqinn** [comment]
> Hadoop and spark are different components, it is better to maintain them separately.

I do not fully agree. In the early days, Hadoop may be special. We have a specific code path to read HADOOP_CONF_DIR. But now Hadoop is an option, as we have other options for storage and scheduling, especially on the cloud or Kubernetes.

Maybe we shall treat it like hive configurations or other third-party components to reduce the maintenance burden and complexity of the deployment.

**Comment by turboFei** [comment]
> Maybe we shall treat it like hive configurations or other third-party components to reduce the maintenance burden and complexity of the deployment.

I believe different companies treat hadoop conf differently.

For ebay, we add conf version for hadoop conf, because it is used by
- public  hadoop client nodes
- private hadoop client nodes
- hadoop service nodes(nn, rm, hms, kyuubi)

image

and we have an RESTful service to download the hdaoop conf and we use soft link to manage them locally.

Recently, we are making migration, from spark3 + hadoop2 to spark3 + hadoop3.

For hadoop2 and hadoop3, their hadoop conf are even different for us.

So to manage the hadoop conf well,  in ebay, we do not want to put the hadoop conf files and spark conf files together.

**Comment by infoankitp** [comment]
I can also help with a use case for this, usually the submission client is on a single environment (Lets say we have it on cloud), and with spark on k8s, we can easily run jobs in different envs like in private Cloud Clusters being submitted from public Cloud. Where we would need diff properties to be passed for the submission client as well as for drivers and executors. This is also a use case where mounting the hadoopConfMap in executors would help in making the task easy to maintain the configs.

**Comment by turboFei** [comment]
> I can also help with a use case for this, usually the submission client is on a single environment (Lets say we have it on cloud), and with spark on k8s, we can easily run jobs in different envs like in private Cloud Clusters being submitted from public Cloud. Where we would need diff properties to be passed for the submission client as well as for drivers and executors. This is also a use case where mounting the hadoopConfMap in executors would help in making the task easy to maintain the configs.


Yes, I think this pr is general for hadoop conf use case, and it does not create more resource because it just use the existing config map.

@yaooqinn @dongjoon-hyun could you help to take another look? Appreciated for your help.

**Review by dongjoon-hyun — APPROVED** [review]
Although this way is a little controversial like @yaooqinn pointed out, I agree with the intention and use-cases and I believe we can allow this way back additionally without many burden or intervention because `spark.kubernetes.executor.hadoopConfigMapName` is still reserved in the code. 

If there is no strong objection from other people, +1 from my side. Please let us know if you still disagree, @yaooqinn .

**Comment by turboFei** [comment]
thanks all !!!

**Comment by yaooqinn** [comment]
thanks, @dongjoon-hyun and @turboFei. Late +1 from my side.

**Comment by pan3793** [comment]
@maomaodev Generally, the executor is supposed to use the delegation token instead of keytab to access kerberized HDFS. I wonder how you set up your Kerberos auth for Spark? And what's the behavior before this PR.

**Comment by turboFei** [comment]
> gentle ping @turboFei In the cluster mode, this merge has caused a new issue in our environment. After the executor mounts the Hadoop conf file and tries to retrieve user information during startup, it fails to start because the conf file contains Kerberos authentication configuration, but the executor does not mount krb5.conf. Have you tested this scenario? 企业微信20250107-163956@2x

How about uploading the krb5.conf?
https://github.com/apache/spark/blob/36d23eff4b4c3a2b8fd301672e532132c96fdd68/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala#L572-L579

**Comment by maomaodev** [comment]
> @maomaodev Generally, the executor is supposed to use the delegation token instead of keytab to access kerberized HDFS. I wonder how do you set up your Kerberos auth for Spark? And what's the behavior before this PR.

1. Yes, the executor does use delegation token, but based on the stacktrace provided above, the code is throwing an exception before adding the delegation token. 
企业微信20250107-171644@2x

2. We use Kyuubi for submission, using the parameter spark.kubernetes.kerberos.krb5.path.
企业微信20250107-173420@2x

3. Prior to this PR, we were using Spark 3.4.2, and the executor did not mount the Hadoop conf during startup. The executor logs showed simple authentication.
企业微信20250107-172445@2x

**Comment by maomaodev** [comment]
> spark.kubernetes.kerberos.krb5.path

The parameter(spark.kubernetes.kerberos.krb5.path) has been set.

**Comment by pan3793** [comment]
@maomaodev I see what happens, the official Spark installs `krb5-user` which generates a default `/etc/krb5.conf` with content
```
[libdefaults]
	default_realm = ATHENA.MIT.EDU
...
```
so that `KerberosUtil.getDefaultRealm` works well, this is not the expected behavior but happens to work. I think the root cause is `spark.kubernetes.kerberos.krb5.path` claims to mount on both driver and executors but actually not.

In short, creating a dummy `/etc/krb5.conf` in your base image should workaround your issue, and the correct solution is mounting `spark.kubernetes.kerberos.krb5.path` to executor pod correctly.

**Comment by maomaodev** [comment]
> @maomaodev I see what happens, the official Spark image installs `krb5-user` which generates a default `/etc/krb5.conf` with content
> 
> ```
> $ docker run --rm apache/spark:3.5.4 cat /etc/krb5.conf
> [libdefaults]
> 	default_realm = ATHENA.MIT.EDU
> ...
> ```
> 
> so that `KerberosUtil.getDefaultRealm` works well, this is not the expected behavior but happens to work. I think the root cause is `spark.kubernetes.kerberos.krb5.path` claims to mount on both driver and executors but actually not.
> 
> In short, creating a dummy `/etc/krb5.conf` in your base image should workaround your issue, and the correct solution is mounting `spark.kubernetes.kerberos.krb5.path` to executor pod correctly.

Yes, creating a dummy /etc/krb5.conf in the base image does work. Is the community planning to fix this issue?

**Comment by pan3793** [comment]
@maomaodev I will prepare a PR soon

**Pull request closed without merging**